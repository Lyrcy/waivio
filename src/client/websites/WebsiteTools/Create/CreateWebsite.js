import React, { useCallback, useEffect } from 'react';
import Helmet from 'react-helmet';
import { FormattedMessage, injectIntl } from 'react-intl';
import PropTypes from 'prop-types';
import { Input, AutoComplete, Checkbox, Form, Button, message } from 'antd';
import { connect } from 'react-redux';
import { debounce, get, isEmpty } from 'lodash';

import Affix from '../../../components/Utils/Affix';
import LeftSidebar from '../../../app/Sidebar/LeftSidebar';
import MobileNavigation from '../../../components/Navigation/MobileNavigation/MobileNavigation';
import validateRules from '../constants/validateRules';
import { checkAvailableDomain, createNewWebsite, getParentDomainList } from '../../websiteActions';
import { getDomainAvailableStatus, getParentDomain, getWebsiteLoading } from '../../../reducers';

import './CreateWebsite.less';

const CreateWebsite = ({
  intl,
  form,
  getDomainList,
  parentDomain,
  checkStatusAvailableDomain,
  availableStatus,
  createWebsite,
  loading,
}) => {
  const { getFieldDecorator, getFieldValue } = form;
  const template = getFieldValue('parent');
  const subDomain = getFieldValue('domain');
  const domainNamesList = Object.keys(parentDomain);
  const available = get(availableStatus, 'status');
  const statusMessageClassList = available ? 'CreateWebsite__available' : 'CreateWebsite__error';
  const domainStatus = useCallback(
    debounce(
      () => checkStatusAvailableDomain(getFieldValue('domain'), parentDomain[template]),
      300,
    ),
    [template],
  );

  useEffect(() => {
    getDomainList();
  }, []);

  useEffect(() => {
    if (subDomain) domainStatus();
  }, [template]);

  const handleSubmit = e => {
    e.preventDefault();
    form.validateFieldsAndScroll((err, values) => {
      if (!err && available)
        createWebsite(values)
          .then(() => form.resetFields())
          .catch(error => message.error(error));
    });
  };
  console.log(availableStatus);
  console.log(loading);
  return (
    <div className="shifted">
      <Helmet>
        <title>
          {intl.formatMessage({
            id: 'create_new_website',
            defaultMessage: 'Create new website',
          })}{' '}
          - Waivio
        </title>
      </Helmet>
      <div className="settings-layout container">
        <Affix className="leftContainer" stickPosition={77}>
          <div className="left">
            <LeftSidebar />
          </div>
        </Affix>
        <div className="center">
          <MobileNavigation />
          <h1>
            <FormattedMessage id="create_new_website" defaultMessage="Create new website" />
          </h1>
          <Form className="CreateWebsite" id="CreateWebsite" onSubmit={handleSubmit}>
            <Form.Item>
              <h3>
                <span className="ant-form-item-required">
                  {intl.formatMessage({
                    id: 'select_website_template',
                    defaultMessage: 'Select website template:',
                  })}
                </span>
              </h3>
              {getFieldDecorator('parent', {
                rules: validateRules.autocomplete,
              })(
                <AutoComplete>
                  {domainNamesList.map(domain => (
                    <AutoComplete.Option key={domain} value={domain}>
                      {domain}
                    </AutoComplete.Option>
                  ))}
                </AutoComplete>,
              )}
            </Form.Item>
            <Form.Item>
              <h3>
                <span className="ant-form-item-required">
                  {intl.formatMessage({
                    id: 'name_website',
                    defaultMessage: 'Specify name for your website:',
                  })}
                </span>
              </h3>
              <div className="CreateWebsite__domain-wrap">
                {getFieldDecorator('domain', {
                  rules: validateRules.domain,
                })(<Input disabled={!template} onInput={domainStatus} />)}
                {template && <span className="CreateWebsite__domain-name">.{template}</span>}
              </div>
            </Form.Item>
            {!isEmpty(availableStatus) && (
              <span className={statusMessageClassList}>
                {intl.formatMessage(availableStatus.intl)}
              </span>
            )}
            <p>
              {intl.formatMessage({
                id: 'info_level_domain_name',
                defaultMessage: 'It will be used as a second level domain name.',
              })}
            </p>
            <Form.Item>
              {getFieldDecorator('politics', {
                rules: validateRules.politics,
              })(
                <Checkbox>
                  <span className="ant-form-item-required">
                    {intl.formatMessage({
                      id: 'website_create_alerts_info',
                      defaultMessage:
                        'I have read and agree to the terms and conditions of the Web Hosting Service Agreement',
                    })}
                  </span>
                </Checkbox>,
              )}
            </Form.Item>
            <Form.Item>
              <Button type="primary" htmlType="submit" loading={loading}>
                {intl.formatMessage({
                  id: 'create_website',
                  defaultMessage: 'Create website',
                })}
              </Button>
            </Form.Item>
          </Form>
        </div>
      </div>
    </div>
  );
};

CreateWebsite.propTypes = {
  intl: PropTypes.shape().isRequired,
  form: PropTypes.shape().isRequired,
  getDomainList: PropTypes.func.isRequired,
  createWebsite: PropTypes.func.isRequired,
  parentDomain: PropTypes.arrayOf(PropTypes.string).isRequired,
  checkStatusAvailableDomain: PropTypes.func.isRequired,
  availableStatus: PropTypes.string,
  loading: PropTypes.bool.isRequired,
};

CreateWebsite.defaultProps = {
  availableStatus: '',
};

export default connect(
  state => ({
    parentDomain: getParentDomain(state),
    availableStatus: getDomainAvailableStatus(state),
    loading: getWebsiteLoading(state),
  }),
  {
    getDomainList: getParentDomainList,
    checkStatusAvailableDomain: checkAvailableDomain,
    createWebsite: createNewWebsite,
  },
)(Form.create()(injectIntl(CreateWebsite)));
