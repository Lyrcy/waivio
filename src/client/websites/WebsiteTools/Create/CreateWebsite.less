@import (reference) '../../../styles/custom.less';

.CreateWebsite {
  &__domain-wrap {
    display: flex;

    .ant-input {
      width: 50%;
      margin-right: 5px;
    }
  }

  &__available {
    color: @green;
  }

  &__error {
    color: @red;
  }

  &__domain-name {
    font-weight: bolder;
  }

  .ant-form-explain {
    margin: 0;
    min-height: auto;
  }

  .ant-select-auto-complete.ant-select .ant-input {
    background: @white;
  }
}

@import 'CreateWebsite-nightmode';
