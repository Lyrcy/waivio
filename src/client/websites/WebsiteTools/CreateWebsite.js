import React from 'react';
import Helmet from 'react-helmet';
import { FormattedMessage, injectIntl } from 'react-intl';
import PropTypes from 'prop-types';

import Affix from '../../components/Utils/Affix';
import LeftSidebar from '../../app/Sidebar/LeftSidebar';
import MobileNavigation from '../../components/Navigation/MobileNavigation/MobileNavigation';

const CreateWebsite = ({ intl }) => (
  <div className="shifted">
    <Helmet>
      <title>
        {intl.formatMessage({
          id: 'create_new_website',
          defaultMessage: 'Create new website',
        })}{' '}
        - Waivio
      </title>
    </Helmet>
    <div className="settings-layout container">
      <Affix className="leftContainer" stickPosition={77}>
        <div className="left">
          <LeftSidebar />
        </div>
      </Affix>
      <div className="center">
        <MobileNavigation />
        <h1>
          <FormattedMessage id="create_new_website" defaultMessage="Create new website" />
        </h1>
        <div className="Settings__section">
          <h3>
            <FormattedMessage
              id="select_website_template"
              defaultMessage="Select website template:"
            />
          </h3>
        </div>
      </div>
    </div>
  </div>
);

CreateWebsite.propTypes = {
  intl: PropTypes.shape().isRequired,
};

export default injectIntl(CreateWebsite);
