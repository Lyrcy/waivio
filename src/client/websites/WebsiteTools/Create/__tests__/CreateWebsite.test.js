import React from 'react';
import {mount} from 'enzyme';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import {IntlProvider} from "react-intl";
import {BrowserRouter as Router} from 'react-router-dom';

import CreateWebsite from "../CreateWebsite";

describe('<CreateWebsite />', () => {
  const mockStore = configureStore();
  const props = {
    form: {
      validateFieldsAndScroll: jest.fn(),
      getFieldValue: jest.fn(),
      getFieldDecorator: jest.fn(() => c => c),
    },
    getDomainList: () => {
    },
    parentDomain: [],
    checkStatusAvailableDomain: () => {
    },
    availableStatus: '',
    createWebsite: () => {
    },
    loading: false,
  };

  const wrapper = mount(
    <Provider store={mockStore()}>
      <Router>
        <IntlProvider locale="en">
          <CreateWebsite {...props} />
        </IntlProvider>
      </Router>
    </Provider>
  );

  it('renders without exploding', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('getDomainList should called when component did mount', () => {
    expect(props.getDomainList).toHaveBeenCalled();
  })

  it('form submit', () => {

    wrapper.find('#CreateWebsite').simulate('submit');
    expect(props.createWebsite).toHaveBeenCalled();
  });
})
