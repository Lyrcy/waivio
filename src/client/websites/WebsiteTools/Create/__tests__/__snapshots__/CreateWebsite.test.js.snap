// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateWebsite /> renders without exploding 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <UNDEFINED
      availableStatus=""
      checkStatusAvailableDomain={[Function]}
      createWebsite={[Function]}
      form={Object {}}
      getDomainList={[MockFunction]}
      loading={false}
      parentDomain={
        Array [
          Object {
            "12345": "parent",
          },
        ]
      }
    />
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <UNDEFINED
        availableStatus=""
        checkStatusAvailableDomain={[Function]}
        createWebsite={[Function]}
        form={Object {}}
        getDomainList={[MockFunction]}
        loading={false}
        parentDomain={
          Array [
            Object {
              "12345": "parent",
            },
          ]
        }
      />,
      "value": Object {
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "subscription": Subscription {
          "handleChangeWrapper": [Function],
          "listeners": Object {
            "notify": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": null,
        },
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "availableStatus": "",
        "checkStatusAvailableDomain": [Function],
        "createWebsite": [Function],
        "form": Object {},
        "getDomainList": [MockFunction],
        "loading": false,
        "parentDomain": Array [
          Object {
            "12345": "parent",
          },
        ],
      },
      "ref": null,
      "rendered": null,
      "type": Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(Form(InjectIntl(CreateWebsite)))",
        "type": [Function],
        Symbol(Symbol.hasInstance): [Function],
      },
    },
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "_calculateChangedBits": null,
          "_context": [Circular],
        },
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": null,
        "_currentValue2": null,
        "_threadCount": 0,
        "displayName": "ReactRedux",
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <UNDEFINED
          availableStatus=""
          checkStatusAvailableDomain={[Function]}
          createWebsite={[Function]}
          form={Object {}}
          getDomainList={[MockFunction]}
          loading={false}
          parentDomain={
            Array [
              Object {
                "12345": "parent",
              },
            ]
          }
        />,
        "value": Object {
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "subscription": Subscription {
            "handleChangeWrapper": [Function],
            "listeners": Object {
              "notify": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": null,
          },
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "availableStatus": "",
          "checkStatusAvailableDomain": [Function],
          "createWebsite": [Function],
          "form": Object {},
          "getDomainList": [MockFunction],
          "loading": false,
          "parentDomain": Array [
            Object {
              "12345": "parent",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(Form(InjectIntl(CreateWebsite)))",
          "type": [Function],
          Symbol(Symbol.hasInstance): [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": Object {
            "$$typeof": Symbol(react.context),
            "_calculateChangedBits": null,
            "_context": [Circular],
          },
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": null,
          "_currentValue2": null,
          "_threadCount": 0,
          "displayName": "ReactRedux",
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
